test/upgrade/main.go

```diff
@@ -391,39 +424,135 @@ func createGameServers(ctx context.Context, gsPath string) {
-					continue
+			newGs := newObj.(*agonesv1.GameServer)
+			if newGs.Status.State == "Error" || newGs.Status.State == "Unhealthy" {
+				gsVersion := newGs.Labels["agonesVersion"]
+				sdkVersion := newGs.Annotations["agones.dev/sdk-version"]
+				log.Printf("Game server %s with binary version %s, and SDK version %s in %s state\n",
+					newGs.Name, gsVersion, sdkVersion, newGs.Status.State)
+
+				// Put failed game servers into the map until it reaches capacity.
+				failedGs[newGs.Name] = gsLog{GameServerVersion: gsVersion, SdkVersion: sdkVersion,
+					GameServerState: string(newGs.Status.State)}
+				if len(failedGs) > acceptedFailures {
+					log.Fatalf("Too many Game Servers in Error or Unhealthy states: %v", failedGs)
 				}
-				gsVersion := newPod.Labels["agonesVersion"]
-				sdkVersion := newPod.Annotations["agones.dev/sdk-version"]
-				log.Printf("%s for pod: %s with game server binary version %s, and SDK version %s", cs.State.Waiting.Reason, newPod.Name, gsVersion, sdkVersion)
-				// Put failed pods into the map until it reaches capacity.
-				failedPods[newPod.Name] = podLog{GameServerVersion: gsVersion, SdkVersion: sdkVersion}
-				if len(failedPods) > acceptedFailures {
-					log.Fatalf("Too many Game Server pods in CrashLoopBackOff: %v", failedPods)
+			}
+		},
+	})
+	if err != nil {
+		log.Fatal("Not able to create AddEventHandler", err)
+	}
+
+	go gsInformer.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, gsInformer.HasSynced) {
+		log.Fatal("Timed out waiting for game server informer cache to sync")
+	}
+}
+
+// watchGameServerEvents watches all `sdk-client-test` containers for BackOff errors. The purpose is
+// to catch ImagePullBackOff and CrashBackOffLoop errors.
+func watchGameServerEvents(kubeClient *kubernetes.Clientset) {
+	stopCh := make(chan struct{})
+	failedPods := make(map[string]int)
+	acceptedFailures := 6 // (6 * 5*time.Second defaultResync) == 30 seconds until failure
+
+	// Filter by Game Server `sdk-client-test` containers
+	containerName := "sdk-client-test"
+	containerPath := "spec.containers{sdk-client-test}"
+	fieldSelector := fields.OneTermEqualSelector("involvedObject.fieldPath", containerPath).String()
+	// First delete previous `sdk-client-test` events, otherwise there will be events from previous runs.
+	_, err := runExecCommand(KubectlCmd, []string{"delete", "events", "--field-selector", fieldSelector}...)
+	if err != nil {
+		log.Fatal("Could not delete `sdk-client-test` events", err)
+	}
+
+	eventOptions := informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+		opts.FieldSelector = fieldSelector
+	})
+	kubeInformerFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, 5*time.Second,
+		informers.WithNamespace("default"), eventOptions)
+	eventInformer := kubeInformerFactory.Core().V1().Events().Informer()
+
+	_, err = eventInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newObj interface{}) {
```

Comment:
Should it be AddFunc?

Reply:
I think this would be `update` since we're watching the event stream of an existing (even if it's not yet fully functioning) resource. Should I add a watch for Add?

Reply:
From adding in some logs to an `AddFunc` and the `UpdateFunc`, it looks like all the events are first captured by `AddFunc`, and then again multiple times by `UpdateFunc` each `defaultResync` which in this case is 5 seconds. So it does seem to make more sense to use `AddFunc`. On that same note, we can actually get rid of the `BackOff` code, as I had made an update to the Game Server template to change the restart policy to "never", so the game server pod will not go into a CrashBackOff loop as one pod can only live once. The container crashing will be caught by the watch for the `Unhealthy` game server.